### ISSUE TITLE
BUG: Numpy object arrays improperly convert python-int's to python-float's. 

### ISSUE THREAD

**Issue by bobbyocean** [numpy/numpy#26818]
### Describe the issue:

It appears during casting operations, numpy will unexpectedly convert large object ints to floats. Why was ONLY array `B` below converted to a float? 

### Reproduce the code example:

```python
import numpy as np
A = np.array([1,1],dtype=object) + [2**62,0]
B = np.array([1,1],dtype=object) + [2**63,0]
C = np.array([1,1],dtype=object) + [2**64,0]
D = np.array([1,1],dtype=object) + [2**63]
E = np.array([1,1],dtype=object) + [2**63,2**63]
print(type(A[0]),A[0]) # <class 'int'> 4611686018427387905
print(type(B[0]),B[0]) # <class 'float'> 9.223372036854776e+18
print(type(C[0]),C[0]) # <class 'int'> 18446744073709551617
print(type(D[0]),D[0]) # <class 'int'> 9223372036854775809
print(type(E[0]),E[0]) # <class 'int'> 9223372036854775809
```

### Error message:

```shell
None.
```

### Python and NumPy Versions:

> import sys, numpy; print(numpy.__version__); print(sys.version)
1.24.3
3.11.4 (main, Jul  5 2023, 14:15:25) [GCC 11.2.0]

### Runtime Environment:

> import numpy; numpy.show_runtime()
[{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2',
                                'AVX512F',
                                'AVX512CD',
                                'AVX512_SKX',
                                'AVX512_CLX',
                                'AVX512_CNL',
                                'AVX512_ICL'],
                      'not_found': ['AVX512_KNL', 'AVX512_KNM']}},
 {'filepath': '/tools/anaconda/Anaconda_2023.07/lib/libmkl_rt.so.2',
  'internal_api': 'mkl',
  'num_threads': 80,
  'prefix': 'libmkl_rt',
  'threading_layer': 'intel',
  'user_api': 'blas',
  'version': '2023.1-Product'},
 {'filepath': '/tools/anaconda/Anaconda_2023.07/lib/libiomp5.so',
  'internal_api': 'openmp',
  'num_threads': 160,
  'prefix': 'libiomp',
  'user_api': 'openmp',
  'version': None}]

### Context for the issue:

Was working on several large combinatorial problems. At some threshold all my arrays converted to floats. Had to dig through many many loops to isolate a specific place where it was happening. I now need to go replace all my numpy casting operations with python for-loops to prevent ending up with floats (as the example code shows, adding scalers appears to not cause this problem).

**Comment by shraf** [comment]
feels like weird-inconsistent behavior, why b,c,d,e not just converted to int128

**Comment by mhvk** [comment]
Let's narrow it down a bit:
```
print(np.array([2**62,0]).dtype)
print(np.array([2**63,0]).dtype)
print(np.array([2**64,0]).dtype)
print(np.array([2**63]).dtype)
print(np.array([2**63,2**63]).dtype)
# gives
int64
float64
object
uint64
uint64
```
I do agree it is somewhat illogical that `uint64` is not also picked for B. Maybe @seberg can advice?

**Comment by seberg** [comment]
I once had a plan that could have allowed it, but then decided it is a lot of trouble for little gain (and others voiced that opinion also).
You could still do something similar if you create a special dtype, but you need to finalize that away, etc.

But, dunno if the whole "int64" or "uint64" (or "intp" on 32 bit systems actually) isn't just as much of a problem anyway.  And promoting int and uint to float64 is also weird everywhere, not just here.

My suspicion is, that it would be fine to at least only remove any uint/float64 possibility and either just go to `object` or (consistency wise better) just raise an error and force the user to resolve.  (it seems unlikely the user doesn't know what they want here)

**Comment by seberg** [comment]
Also xref gh-20905, I would be surprised if there isn't an open issue for this as well, but didn't find it in a quick search.

For this issue, a duplicate is for example gh-19146.

**Comment by mhvk** [comment]
Thanks for the links. @bobbyocean - in your case it is perhaps the addition that was most surprising, that maybe the list should be converted to object array in the first place?  If so, this is a somewhat different issue; if not, we probably should close it as a duplicate of gh-19146.

**Comment by bobbyocean** [comment]
@mhvk Oh, wow. Thanks for the simplified code. I didn't realize that lists of python ints also sometimes covert to floats, directly without adding anything. 

Although the underlying cause is the same reasoning. I do feel this is technically addressing a slightly different inconsistent behavior from https://github.com/numpy/numpy/issues/19146. Namely, that an array's defined "dtype" is not used to make a determination about how to perform arithmetic; but to the end user they would think that is the case. 

```
A = np.array([1,2],dtype=object) # here the user thinks they are creating an array to work with python dynamic ints
A += [8,9] # here the user thinks they are adding python dynamic ints to the array and the array will still have python dynamic ints (which it still does)
print(type(A[0])) # still a python dynamic int; not numpy int64 or uint64, etc. 
```
vs
```
A = np.array([1,2],dtype=object) # here the user thinks they are creating an array to work with python dynamic ints
A += [2**63,0] # here the user thinks they are adding python dynamic ints to an array of python dynamic ints; surprisingly there original array was converted to python floats
print(type(A[0]))
```
and of course this behavior changes when ```A += 2**63``` (still python dynamic ints; not int64, uint64, not floats) and ```A += [2**63,2**63]``` (still python dynamic ints; not int64, uint64, not floats). 

The point is that a user creates an array for dynamic int processing. Nearly every operation results in the array staying as python dynamic ints. Unsuspectedly adding the wrong list of python dynamic ints to the array, converts the entire array to python floats. 

This creates other unexpected results as well. For example, 

```
np.array([7,'a'],object)*[2**62,2] # works perfectly fine and makes sense, array stays as an object array; both the 7 and the 'a' remain the same python type they started as, both behaved as expected
np.array([7,'a'],object)*[2**63,2] # raises an error 
```

**Comment by mhvk** [comment]
Indeed, the basic problem is that a list gets converted to an array without taking into account what kind of operation it is involved in. While this is easy to state, fixing it may be quite a bit harder  (it may be possible to change this behaviour by diving into the ufunc machinery, but that needs finding someone sufficiently interested to take that on...)  

Overall, `object` arrays are somewhat strange beasts, which do not really fit the regular `ndarray` mold so well (I think quite a few array alternatives do not support them at all), so perhaps the lesson here is that one has to be careful in explicitly converting other items to object arrays oneself.

**Comment by seberg** [comment]
Just for those curious, I am not sure it's worth trying to do much about it because there are bound to be odd cases here.  It is very possible for ufuncs, but you lose quickly for non-ufuncs because at that point you calling `np.asarray()` first will change behavior!


It is possible to do this.  Not even impossibly hard: my first prototype for new dtypes and ufuncs even implemented it (actually with an additional subtlety to be even clearer since I split finding DType and dtype in two allowing to properly define more things).  In fact, we effectively already have to do this for Python scalars.

The trick, is that converting to an array consists of (if not an array):
* discovery of the array-like dtype
* creation of the actual array from the elements.

So within ufuncs, it is very possible to rearrange:
1. convert array:
   * discover dtype
   * create new array
2. Find the loop dtypes
3. Do the operation

To split up the first part so that it is:
* Discover dtypes
* Find the loop dtypes
* Convert any non-array already knowing the correct dtype.

This would subtly change behavior in some cases, but in practice mostly with object dtypes.  You would also still get the incorrect `float64` intermediate (but throw it away).

**Comment by randomseed42** [comment]
Just observed same behavior when I'm doing some data type conversion between numy and sql, of course this edge case is found during unittest only, I guess this might be worthy to correct. Below is the minimal code to explain, and I assume it is better to be consistency between uint32 and uint64. Thanks!

```python
>>> import numpy as np
>>>
>>> np.array([0, np.iinfo(np.uint32).max])
array([         0, 4294967295])
>>>
>>> np.array([0, np.iinfo(np.uint64).max])
array([0.00000000e+00, 1.84467441e+19])
>>>
>>> np.array([0, np.iinfo(np.uint64).max], dtype=np.uint64)
array([                   0, 18446744073709551615], dtype=uint64)
>>>
>>> np.array([0, np.iinfo(np.uint64).max]).dtype
dtype('float64')
>>>
```
